### ISSUE TITLE
Stage offset not updated when non-crafty DOM changes

### ISSUE THREAD

**Issue by benbarefield** [craftyjs/Crafty#452]
Bad title, but I didn't know how to best summarize this problem.
The page that my game is on currently looks like this and I use Canvas for my game elements:

``` html
<div style="position:relative;left:-50%;float:right;">
    <div style="position:relative;left:50%">                
        <div id="cr-stage" style="float:left;position:relative">
            <div id="nameInput" style="position:relative;left:400;top:490;z-index:5">
                <label for="name" style="margin-right:5px">Your name:</label>
                <input id="name" type="text" style="width:200"></input>
            </div>
        </div>
        <div style="float:left;position:relative;margin-left:5px">
            Top Times:
            <ul id="topTimes" style="list-style-type:none;padding:0;"></ul>
        </div>
        <script type="text/template" id="topTime-template">
            <div class="topTimeEntry">
                <div class="playerName" style="float:left;padding-right:25px">
                    <%= playerName %>
                </div>
                <div class="topTime" style="float:right">
                    <%= timeSurvived %>
                </div>
                <div style="clear:both"></div>
            </div>
        </script>
    </div>
</div>
```

The topTimes ul gets filled with a name and a time in the format shown in the topTime-template. The styling in the top and second level divs is to center the game div and the top times div on the screen. This means that when the topTimes ul fills, it pushes the cr-stage div to the left. 
The problem seems to be that even though the graphics all get pushed left properly, the translation for the mouse coordinates when clicking do not, so you have to click a button in the position it would be in if the topTimes ul were empty (the cr-stage div would be in the center of the page), not where it shows up on the screen.

**Comment by benbarefield** [comment]
After diving into this a bit more, I'm thinking that there isn't a real solution to this and I just need to rework my page layout.
I'm going to leave this open in case anyone has more thoughts.

**Comment by starwed** [comment]
Could you provide an html snippet that directly reproduces this issue, rather than just the template?  That would make it easier to see what's going on.

**Comment by benbarefield** [comment]
Sure - sorry I didn't include this in the first go around, the #topTimes ul has to get filled in _after_ Crafty.init is called. Here is a fiddle:
http://jsfiddle.net/zyrtY/
You can see the problem by trying to click on the left side of the red button, and nothing happens. If you click to the right of the red button, you'll get the alert.

**Comment by starwed** [comment]
Ah, it's because the viewport's offset is calculated once when it is initialized.  

Calling `Crafty.viewport.reload();` after you manipulate the DOM will fix the problem.  This is done automatically when the window is resized, but unfortunately I don't see a way for Crafty to get notified when the DOM has been changed.

**Comment by mucaho** [comment]
It is possible to get notified when the DOM changes:
- [Mutation Events](https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Mutation_events) have best cross-browser support, but terribad performance (DOM Changes up to 7x times more costly)
- [Mutation Observers](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) are supported only by the newest browsers, but have good performance

On the other hand, I think it would be better to add more documentation for `Crafty.viewport.reload` and let the user call it, if he needs too. I will make a new PR about this.

**Comment by mucaho** [comment]
Are there other scenarios where the DOM could change by a 3rd party without the user's knowledge about it? In that case we could listen to mutations and trigger a reload accordingly.

**Comment by mucaho** [comment]
@starwed Can we close this issue? 
Since events have unacceptable performance costs and observers are not supported cross-browser we do not have a meaningful way to detect DOM changes. Hence the user should manually call viewport.reload.

**Comment by starwed** [comment]
Well, we haven't resolved it, and if someone came up with a clever enough method I'd want to land it -- so it's not a WONTFIX in principle. 

Plus, it might be useful for someone who encounters the same issue!

**Comment by sbyrnes321** [comment]
Sorry for the stupid question: We want to avoid checking the DOM (e.g. calling `Crafty.stage.elem.getBoundingClientRect()`) each frame or each mouse-event, because that might lead to performance problems -- is that correct?

**Comment by starwed** [comment]
Yes, it would have some performance impact and wouldn't actually be useful for most games.

I haven't checked how much of an impact it would have, though.  And maybe we could let folk opt-in to the check.
